src/main/java/org/elasticsearch/search/aggregations/bucket/DeferringBucketCollector.java

```diff
@@ -138,6 +138,9 @@ public void replay(long... selectedBuckets) throws IOException {
         this.selectedBuckets = hash;
 
         collector.preCollection();
+        if (collector.needsScores()) {
+            throw new ElasticsearchIllegalStateException("Cannot defer if scores are needed");
+        }
```

Comment:
This is already checked earlier in TermsAggregator.shouldDefer. I see it as an invariant check.